sharedbuffers/mapped_struct.py

```diff
@@ -1112,20 +1112,43 @@ def Proxy(self):
     
     def __init__(self, slot_types, alignment = 8, pack_buffer_size = 65536, packer_cache = None, unpacker_cache = None,
             max_pack_buffer_size = None):
-        if packer_cache is None:
-            packer_cache = Cache(256)
-        if unpacker_cache is None:
-            unpacker_cache = Cache(256)
-        self.slot_types = self._map_types(slot_types)
-        self.pack_buffer_size = pack_buffer_size
-        if max_pack_buffer_size is not None:
-            self.max_pack_buffer_size = max_pack_buffer_size
-        else:
-            self.max_pack_buffer_size = max(128<<20, max(pack_buffer_size, min(pack_buffer_size * 2, 0x7FFFFFFF)))
-        self.alignment = alignment
-        self.packer_cache = packer_cache
-        self.unpacker_cache = unpacker_cache
-        self.init()
+        self.init(
+            self._map_types(slot_types),
+            packer_cache = packer_cache, unpacker_cache = unpacker_cache, alignment = alignment,
+            pack_buffer_size = pack_buffer_size)
+
+    def __reduce__(self):
+        return (type(self), (self.slot_types,), self.__getstate__())
+
+    def __getstate__(self):
+        return dict(
+            slot_types = self.slot_types,
+            slot_keys = self.slot_keys,
+            alignment = self.alignment,
+        )
+
+    def __setstate__(self, state):
+        # These are a safety hazard (they could cause OOM if corrupted), so ignore them if present
+        # Not needed for unpacking anyway
+        state.pop('pack_buffer_size', None)
+        state.pop('max_pack_buffer_size', None)
+
+        self.init(**state)
+
+    @cython.locals(other_schema = 'Schema')
+    def compatible(self, other):
+        if not isinstance(other, Schema):
+            return False
+
+        other_schema = other
+        if self.slot_keys != other_schema.slot_keys or self.alignment != other_schema.alignment:
+            return False
+
+        for k in self.slot_keys:
+            if self.slot_types[k] is not other_schema.slot_types[k]:
+                return False
```

Comment:
Using all there is complex. All takes a generator/list, and that slot_types there is a private attribute of the Schema only accessible through Cython. I'm pretty sure to make it work it would either be ugly (making a type-annotated generator that would take 4x the line count) or inefficient (building a list)